### PR TITLE
Enhanced API testing

### PR DESCRIPTION
**Pull request by cachedout**
This enhances the test framework to support recent changes to the API.

PRs which this covers:

- [x] https://github.com/elastic/logstash/pull/10885
- [x] https://github.com/elastic/logstash/pull/10853
- [x] https://github.com/elastic/logstash/pull/10589
- [x] https://github.com/elastic/logstash/pull/10576
- [x] https://github.com/elastic/logstash/pull/10561

It also covers a few missing cases found along the way.

Refs: https://github.com/elastic/kibana/issues/33706

### CODE REVIEW

**Comment by cachedout** [comment]
Test failure appears unrelated to these changes.

**Comment by cachedout** [comment]
This is ready for review now. We might want to consider testing around the new metadata system for plugins but that is a task that may need to be discussed in more depth with the Logstash team.

**Comment by ycombinator** [comment]
Since this PR doesn't introduce new functionality but is Ruby code involving writing tests for existing functionality, I'll defer to someone from the Logstash core team for the review.

**Comment by cachedout** [comment]
Hi @jsvd . We are tracking this for 7.4 right now. Do you think you might be able to look at this in the next few weeks?

File: logstash-core/spec/logstash/api/commands/stats_spec.rb [hunk 1]
```diff
@@ -35,18 +73,93 @@
...
+      :successes,
+      :failures,
+      )
+    end
   end
+
+  describe "pipeline stats" do
+    let(:report_method) { :pipeline }
+
+    it "returns pipelines information" do
```
Comment (jsvd) [inline logstash-core/spec/logstash/api/commands/stats_spec.rb]: I suggest breaking this large test into smaller checks, e.g:

```ruby
    it "returns information on existing pipeline" do
      expect(report.keys).to include(:main)
    end
    context "for each pipeline" do
      it "returns information on pipeline" do
        expect(report[:main].keys).to include(
          :events,
          :plugins,
          :reloads,
          :queue,
        )
      end
      it "returns event information" do
        expect(report[:main][:events].keys).to include(
          :in,
          :filtered,
          :duration_in_millis,
          :out,
          :queue_push_duration_in_millis
        )
      end
    end
```

File: logstash-core/spec/logstash/api/commands/node_spec.rb [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+      )
+    expect(report[:mem].keys).to include(
+      :heap_init_in_bytes,
+      :heap_max_in_bytes,
+      :non_heap_init_in_bytes,
+      :non_heap_max_in_bytes
+    )
+    end
+  end
+  describe "#decorate_with_cluster_uuid does not mutate" do
```
Comment (jsvd) [inline logstash-core/spec/logstash/api/commands/node_spec.rb]: not sure I understand the goal of this test. is that `decorate_with_cluster_uuid` doesn't  add/change the structure of the report?

Comment (cachedout) [reply logstash-core/spec/logstash/api/commands/node_spec.rb]: That was the idea but when looking at this again, I don't think it's doing an effective job of that so I just removed it.

File: logstash-core/spec/logstash/api/commands/stats_spec.rb (hunk 1 above)
Comment (cachedout) [reply logstash-core/spec/logstash/api/commands/stats_spec.rb]: Thanks. Suggestion applied.

**Pull request merged**